src/Engine/EmitMSIL/Replication.cs

```diff
@@ -319,43 +355,180 @@ private static List<List<ReplicationGuide>> ConstructRepGuides(string[][] replic
+            MSILRuntimeCore runtimeCore,
+            List<CLRFunctionEndPoint> feps,
+            List<CLRStackValue> argumentsList)
+        {
+            /*TODO_MSIL: figure out the this ptr context
+            Validity.Assert(thisPtr.IsNull || thisPtr.IsPointer,
+                            "this pointer wasn't null or a pointer. {89635B06-AD53-4170-ADA5-065EB2AE5858}");
+            */
+
+            var classId = feps.FirstOrDefault().procedureNode.ClassID;
+
+            // We have multiple possible scopes for the function call:
+            // 1. Static method call - no this pointer
+            // ex: ClassA.Method();
+            //    Hidden static methods generate multiple feps.
+            //    We do not need to check actually if the method has the "IsHideBySig" (https://docs.microsoft.com/en-us/dotnet/api/system.reflection.methodbase.ishidebysig)
+            //    because static methods can only be hidden.
+            // 2. Method call from an instance of a class - valid this pointer.
+            // ex: classAInstance.method();
+            // 3. Function from the global scope - no this pointer and no class scope.
+            // ex: SomeGlobalFunction();
+            //
+            // All 3 cases will run through the same matching steps.
+
+            // A static function call has an invalid this pointer and a valid class scope;
+            // TODO_MSIL: figure out the this ptr context
+            bool isValidStaticFuncCall = /*!thisPtr.IsNull && !thisPtr.IsPointer &&*/ classId != Constants.kInvalidIndex;
+            if (isValidStaticFuncCall)
+            {
+                return feps.FirstOrDefault();
+            }
+
+            // TODO_MSIL: figure out the this ptr context
+            int typeID = /*thisPtr.TypeUID;*/Constants.kInvalidIndex;
+
+            // Try to match with feps belonging to the class scope (most derived class should have priority).
+            // In this case we simply select the function that belongs to the calling class.
+            // The assumption here is that all function end points in "feps" have already been checked that they have the same signature.
+            IEnumerable<CLRFunctionEndPoint> exactFeps = feps.Where(x => x.procedureNode.ClassID == typeID);
+            if (exactFeps.Count() == 1)
+            {
+                return exactFeps.First();
+            }
+
+            //Walk the class tree structure to find the method
+            while (runtimeCore.ClassTable.ClassNodes[typeID].Base != Constants.kInvalidIndex)
+            {
+                typeID = runtimeCore.ClassTable.ClassNodes[typeID].Base;
+
+                foreach (var fep in feps)
+                    if (fep.procedureNode.ClassID == typeID)
+                        return fep;
+            }
+
+            //We weren't able to distinguish based on class hiearchy, try to sepearete based on array ranking
+            List<int> numberOfArbitraryRanks = new List<int>();
```

Comment:
var?

Reply:
do you think it could make sense performance wise to allocate these lists at the start of the while loop and reset them in each for loop?

Reply:
Maybe...I do not want to do such fine grain refactor (this is copy-pasted code)
I propose we refactor later when we get further ahead